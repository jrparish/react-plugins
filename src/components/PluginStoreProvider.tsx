import React, { useMemo } from 'react';
import PluginStore from '../utils/store';

type ProviderContextType = {
  store: PluginStore;
};

export const ProviderContext = React.createContext<ProviderContextType>(
  {} as ProviderContextType
);

type PluginStoreProviderProps = ProviderContextType;

const PluginStoreProvider: React.FC<PluginStoreProviderProps> = ({
  children,
  store
}) => {
  const contextValues = useMemo(
    () => ({
      store
    }),
    [store]
  );

  return (
    <ProviderContext.Provider value={contextValues}>
      {children}
    </ProviderContext.Provider>
  );
};

export default PluginStoreProvider;
