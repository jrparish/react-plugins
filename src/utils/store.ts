import { ComponentType } from 'react';
import { PluginStoreType } from '../types';

export class PluginStore {
  private sections: PluginStoreType = {};
  private listeners: Array<(sections: PluginStoreType) => void> = [];

  subscribe(listener: (sections: PluginStoreType) => void): () => void {
    this.listeners.push(listener);

    // return unsubscribe callback
    return (): void => {
      this.listeners = this.listeners.filter((current) => current !== listener);
    };
  }

  registerPlugin<P>(
    section: string,
    component: ComponentType<P>,
    name: string,
    priority = 0,
    props?: P
  ): void {
    const pluginStore = this.sections[section] || [];

    // if we find a component with the given name, replace it
    if (pluginStore.findIndex((current) => current.name === name) !== -1) {
      this.sections[section] = pluginStore.map((current) =>
        current.name === name
          ? {
              component,
              priority,
              name,
              props
            }
          : current
      );
    } else {
      this.sections[section] = [
        ...pluginStore,
        { component, priority, name, props }
      ];
    }

    this.listeners.forEach((listener) => listener(this.sections));
  }

  removePlugin(section: string, name: string): void {
    const pluginStore = this.sections[section] || [];

    this.sections[section] = pluginStore.filter(
      (current) => current.name !== name
    );

    this.listeners.forEach((listener) => listener(this.sections));
  }

  getPluginsForSection(section: string) {
    const pluginStore = this.sections[section];

    if (!pluginStore || pluginStore.length < 1) {
      return [];
    }

    return pluginStore.sort((a, b) => a.priority - b.priority).map((a) => a);
  }
}
