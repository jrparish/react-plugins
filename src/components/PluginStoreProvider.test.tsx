import React from 'react';
import { render } from '@testing-library/react';
import { PluginStoreProvider } from './PluginStoreProvider';
import { PluginStore } from '../utils/store';

describe('PluginStoreProvider', () => {
  it('Renders', () => {
    const { asFragment } = render(
      <PluginStoreProvider store={new PluginStore()} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
