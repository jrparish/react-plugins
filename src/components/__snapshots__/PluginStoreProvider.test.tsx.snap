// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginStoreProvider Renders 1`] = `<DocumentFragment />`;
